### ISSUE TITLE
Converting from TensorFlow model input to MHLO

### ISSUE THREAD

**Issue by Jackwin** [tensorflow/tensorflow#52887]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.4.3
- Are you willing to contribute it (Yes/No):
Yes，I would like to but have no ideas

**Describe the feature and the current behavior/state.**
Have no guide on how to convert a TF model to MLIR, and hope that build up a doc that shows how to make the conversion. Meanwhile, it would be better if there are more tutorials about TF and MLIR. 

**Will this change the current api? How?**
Have no ideas

**Who will benefit with this feature?**
The people who would like to use TensorFlow as the frontend in MLIR. It will enhance the combination of
the TensorFlow eco-system and MLIR eco-system.

**Any Other info.**

**Comment by jpienaar** [comment]
Thanks for filing this, to add context from origin thread, the request was for "standalone infrastructure that imports the TF model and exports the MHLO".  I'll refrain this issue to be scoped to that and assign to bridge team for comment.

I think there are 2 parts here 1) documenting how to perform the conversion (this has been requested in a couple of threads) and 2) the standalone tool.

**Comment by Jackwin** [comment]
Thank Jacques for the refinement. The MLIR thread is repeated [here](https://llvm.discourse.group/t/convert-tensorflow-tf-lite-model-to-mhlo-linalg-standalone/4611/2), which maybe help the guys who are interested in TF and MLIR, hopefully.

**Comment by paynecl** [comment]
Hi Chunjie.  I work with Jacques on the part of the TF stack (bridge team) that can do this conversion.  We are looking at making a standalone tool for this.

**Comment by Jackwin** [comment]
@paynecl Quite excited to get this news . Looking forward to this feature. Thanks ahead:)

**Comment by smit-hinsu** [comment]
Sorry for the late response.

If I understand correctly, you want to convert a saved model loaded by --savedmodel-objectgraph-to-mlir to MHLO? You could do so by 

``` tf-opt out.mlir -tf-standard-pipeline | tf-opt -tf-to-hlo-pipeline ```

If you want to do this in a programmatic way, then you could use CreateTFStandardPipeline and createConvertToMHLOPass.
 
We will add a documentation covering different use-cases.

Let us know if this doesn't address your use-case.

**Comment by jpienaar** [comment]
This may run into saved model dialect remaining. Is there perhaps an open SavedModel example you are looking at @Jackwin ? That could allow ensuring the instructions work.

**Comment by Jackwin** [comment]
Hi @smit-hinsu & @jpienaar 
I randomly picked a saved model named **faster_rcnn_resnet50_v1_640x640** from the [tensorhub](https://tfhub.dev/tensorflow/faster_rcnn/resnet50_v1_640x640/1) in TF version 2.0.

> If I understand correctly, you want to convert a saved model loaded by --savedmodel-objectgraph-to-mlir to MHLO?

Yes. Have tried the following instructions to translate a saved model to mlir, and got the TF excecute dialect.
```
tf-mlir-translate --savedmodel-objectgraph-to-mlir \
--tf-savedmodel-exported-names=predict  \
./faster_rcnn_resnet50_v1_640x640 \
-o faster_rcnn_resnet50_v1_640x640.mlir
```
When converting from TF execute dialect to MHLO, it seems not to work since the output still keeps the TF execute dialect. In this process, several passes have been inserted in `tf-opt` without any effect, such as `--tf-to-hlo-pipeline`, `-tf-saved-model-optimize-global-tensors`, and `-tf-saved-model-optimize-global-tensors`.  Hope that you could provide an example.

> If you want to do this in a programmatic way, then you could use CreateTFStandardPipeline and createConvertToMHLOPass.

Not familiar with these two passes. It would be great if any docs could explain them.

In summary, my concerns are
1. what passes should be picked to complete the translation from a saved model to MHLO?
2. As the saved model usually contains the data, is there any method of getting rid of the data, and just focusing on the graph expression in MLIR? Because the data size in the saved model is quite huge, not friendly to read the MLIR.

Thanks.

**Comment by Jackwin** [comment]
Hi folks, any update about this issue? Thanks

**Comment by smit-hinsu** [comment]
We need to use `--savedmodel-signaturedefs-to-mlir` instead of `--tf-savedmodel-exported-names=predict` for TF v1 models.

Currently, there is no command line method to freeze variables. So, we need to make use of the TFLite Converter with a slight code change. Just remove the block from L247 to L265 [here.](https://github.com/tensorflow/tensorflow/blob/fec390fc795df46555aa79ab89146c3cc3769a1f/tensorflow/compiler/mlir/lite/tf_to_tfl_flatbuffer.cc#L247)

After that the model can be imported in TF dialect with:

```tf_tfl_translate -savedmodel-signaturedefs-to-mlir  --tf-savedmodel-exported-names=serving_default -o ~/Downloads/resnet/translate_tflite_input.mlir  ~/Downloads/resnet/```

Then, we will need to change shape type of argument for @serving_default function from tensor<?x?x?x?xf32> to tensor<?x224x224x3xf32> and similar change for the op that is using this argument.

After that we could run the following command to convert it to mhlo.

```tf-opt -tf-to-hlo-pipeline  -o ~/Downloads/resnet/translate_mhlo.mlir ~/Downloads/resnet/translate_tf.mlir```

Regarding getting rid of the data, there is a mechanism to use splat constants but that is not available for saved model. See [here](https://github.com/tensorflow/tensorflow/blob/ebe7d691f706e4d40a7569fec719a7d1e7404073/tensorflow/compiler/mlir/tensorflow/translate/tf_mlir_translate_registration.cc#L58)

What is the use-case you are trying? Let us know if you have any other questions.

**Comment by tilakrayal** [comment]
@Jackwin ,
Could you please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/52887#issuecomment-1028346791) and let us know if it the issue still persists. Thank you!
